dynamics/orbital_dynamics.py

```diff
@@ -108,3 +122,174 @@ def f_jac(x: np.ndarray, dt: float) -> np.ndarray:
+            [
+                factor * r[0] * (5 * (r[2] ** 2) / r_norm**2 - 1),
+                factor * r[1] * (5 * (r[2] ** 2) / r_norm**2 - 1),
+                factor * r[2] * (5 * (r[2] ** 2) / r_norm**2 - 3),
+            ]
+        )
+        updated_a = base_derivative[3:] + a_J2 + a_drag
+        return np.concatenate([base_derivative[:3], updated_a])
+
+    return wrapper
+
+
+def second_order_effects_jac(func: Callable[[np.ndarray], np.ndarray]) -> Callable[[np.ndarray], np.ndarray]:
+    """
+    The decorator function for computing the Jacobian of the state derivative with second order effects
+    (drag and J2 perturbations).
+    :param func: The state derivative function to be decorated.
+
+    :return: A wrapper function that computes the Jacobian of the state derivative with second order effects.
+    """
+
+    def wrapper(
+        x: np.ndarray,
+        config: dict,
+        data_manager: ODSimulationDataManager,
+    ) -> np.ndarray:
+        """
+        The wrapper function that computes the Jacobian of the state derivative with second order effects.
+        :param x: A numpy array of shape (6,) containing the current state (position and velocity).
+        :param config: The configuration dictionary.
+        :param data_manager: The ODSimulationDataManager instance.
+
+        :return: A numpy array of shape (6, 6) containing the Jacobian of the state derivative with
+        second order effects.
+        """
+        # Extract parameters from the config dictionary
+        base_jacobian = func(x)
+        # pylint: disable=invalid-name
+        CD = config["satellite"]["Cd"]
+        AREA = config["satellite"]["area"]
+        MASS = config["satellite"]["mass"]
+        latest_epoch = data_manager.latest_epoch
+
+        # Compute drag
+        density = density_harris_priester(x=x, epoch=latest_epoch)
+        v = x[3:]
+        v_norm = np.linalg.norm(v)
+        if v_norm == 0:
+            da_drag_dv = np.zeros((3, 3))
+
+        # Compute J2 either using autodiff or manually
+        # j2da_dv = j2_derivative(x[:3])
+        daj2auto_dr = j2_jacobian_auto(x[:3])
+
+        F = -0.5 * density * CD * AREA / MASS
+        da_drag_dv = F * ((np.eye(3) / v_norm) - np.outer(v, v) / v_norm**3)
+
+        da_dr = base_jacobian[3:, :3] + daj2auto_dr
+        da_dv = da_drag_dv
+        # da_dv = np.zeros((3, 3))
+
+        return np.block([[base_jacobian[0:3, 0:3], base_jacobian[0:3, 3:]], [da_dr, da_dv]])
+
+    return wrapper
+
+
+@second_order_effects
+def state_derivative_full(x: np.ndarray) -> np.ndarray:
+    """
+    State derivative function with second order effects (drag and J2 perturbations).
+    :param x: A numpy array of shape (6,) containing the current state (position and velocity).
+
+    :return: A numpy array of shape (6,) containing the state derivative.
+    """
+    return state_derivative(x)
+
+
+def f_full(
+    x: np.ndarray, config: dict, data_manager: ODSimulationDataManager, dt: float
```

Comment:
I think it would make more sense to input the current epoch directly rather than the `ODSimulationDataManager`, since that seems to be the only thing its used for. I think its cleaner for the dynamics to not even have knowledge of that class.